packages/expo-router/src/useScreens.tsx

```diff
@@ -219,30 +220,80 @@ export function getQualifiedRouteComponent(value: RouteNode) {
+      include.set(segment.name, segment);
+    }
   }
-  return ({ params }: { params?: Record<string, any> }) => {
-    const getPreferredId = (segment: DynamicConvention) => {
-      // Params can be undefined when there are no params in the route.
-      const preferredId = params?.[segment.name];
-      // If the route has a dynamic segment, use the matching parameter
-      // as the screen id. This enables pushing a screen like `/[user]` multiple times
-      // when the user is different.
-      if (preferredId) {
-        if (!Array.isArray(preferredId)) {
-          return preferredId;
-        } else if (preferredId.length) {
-          // Deep dynamic routes will return as an array, so we'll join them to create a
-          // fully qualified string.
-          return preferredId.join('/');
+
+  /**
+   * Child routes IDs are a combination of their dynamic segments and the search parameters
+   * As search parameters can be anything, we build an exclude list of its parents dynamic segments.
+   **/
+  if (route.children?.length === 0) {
+    exclude.add('screen');
+    exclude.add('params');
+
+    const contextDynamic = generateDynamic(route.contextKey);
+
+    if (contextDynamic) {
+      for (const segment of contextDynamic) {
+        if (!include.has(segment.name)) {
+          exclude.add(segment.name);
         }
-        // Empty arrays...
       }
-      return segment.deep ? `[...${segment.name}]` : `[${segment.name}]`;
-    };
-    return route.dynamic?.map((segment) => getPreferredId(segment)).join('/');
+    }
+  }
+
+  return ({ params = {} } = {} as { params?: Record<string, any> }) => {
+    let segments: string[] = [];
+
+    const unprocessedParams = new Set(Object.keys(params));
+
+    for (const dynamic of include.values()) {
+      unprocessedParams.delete(dynamic.name);
+      const value = params?.[dynamic.name];
+      if (Array.isArray(value) && value.length > 0) {
+        // If we are an array with a value
+        segments.push(value.join('/'));
+      } else if (value && !Array.isArray(value)) {
+        // If we have a value and not an empty array
+        segments.push(value);
+      } else if (dynamic.deep) {
+        segments.push(`[...${dynamic.name}]`);
+      } else {
+        segments.push(`[${dynamic.name}]`);
+      }
+    }
+
+    let id = segments.join('/');
+
+    segments = [];
+    if (route.children?.length === 0) {
+      for (const key of unprocessedParams) {
+        if (exclude.has(key)) {
+          continue;
+        }
+        segments.push(`${key}=${params[key]}`);
+      }
+    }
+
+    if (segments.length) {
+      id = `${id}?${segments.join('&')}`;
+    }
+
+    /**
+     * We should always return a truthy value, failing to do so will cause React Navigation to
+     * fall back to `key` based navigation. This is an issue for search parameters where are either
+     * part of the key or the id.
+     */
+    return id || route.contextKey;
```

Comment:
We now always return an ID value. This prevents React Navigation from falling back to `name` & `key` based matching, where it may match on a correct `key`, but this screen may have had parameters.